### PR TITLE
Add denunciations in block header

### PR DESCRIPTION
**Pull request by sydhds**
* [ ] document all added functions
* [ ] try in sandbox /simulation/labnet
* [ ] unit tests on the added/changed features
  * [ ] make tests compile
  * [ ] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

File: massa-execution-worker/src/active_history.rs [hunk 1]
```diff
@@ -60,7 +60,7 @@ impl ActiveHistory {
         HistorySearchResult::NoInfo
     }
 
-    /// Lazily query (from end to beginning) the active list of proccessed denunciations.
+    /// Lazily query (from end to beginning) the active list of processed denunciations.
```
Comment (damip) [inline massa-execution-worker/src/active_history.rs:63]: processed => executed

File: massa-execution-worker/src/execution.rs [hunk 2]
```diff
@@ -1527,6 +1527,25 @@ impl ExecutionState {
...
+    /// Check if a denunciation has been executed given a `DenunciationIndex`
+    pub fn is_denunciation_unexecuted(&self, denunciation_index: &DenunciationIndex) -> bool {
+        // check active history
+        let history = self.active_history.read();
+
+        if matches!(
+            history.fetch_processed_de(denunciation_index),
+            HistorySearchResult::Present(())
+        ) {
+            return true;
```
Comment (damip) [inline massa-execution-worker/src/execution.rs:1539]: I think "unexecuted" should be changed to "executed" everywhere because it causes confusion.
Here for example it caused a `return true` but the denunc WAS executed

File: massa-models/src/denunciation.rs [hunk 3]
```diff
@@ -797,6 +798,31 @@ impl From<&DenunciationPrecursor> for DenunciationIndex {
...
+            (self_de, other_de) => {
+                // Note: does not specify an order for Block header de idx versus Endorsement de idx
+                //       but this should not be an issue
+                self_de.get_slot().cmp(other_de.get_slot())
+            }
+        }
+    }
+}
+
+impl PartialOrd for DenunciationIndex {
```
Comment (damip) [inline massa-models/src/denunciation.rs:820]: Here same:

```
fn partial_cmp(&self, other: &Self) -> Ordering
    (slot, index).partial_cmp(&(other.slot, other.index))
}
```

File: massa-pool-worker/src/denunciation_pool.rs [hunk 4]
```diff
@@ -177,15 +186,32 @@ impl DenunciationPool {
...
 
     // In next PR
-    /*
-    pub fn get_block_denunciations(
-        &self,
-        slot: &Slot,
-        target_block: &BlockId,
-    ) -> Vec<DenunciationId> {
-        todo!()
+    pub fn get_block_denunciations(&self, target_slot: &Slot) -> Vec<Denunciation> {
```
Comment (damip) [inline massa-pool-worker/src/denunciation_pool.rs:189]: TODO doc

File: massa-execution-worker/src/active_history.rs (hunk 1 above)
Comment (sydhds) [reply massa-execution-worker/src/active_history.rs]: done

File: massa-execution-worker/src/execution.rs (hunk 2 above)
Comment (sydhds) [reply massa-execution-worker/src/execution.rs:1539]: Yes, always avoid double negation :)

File: massa-models/src/denunciation.rs [hunk 5]
```diff
@@ -797,6 +798,31 @@ impl From<&DenunciationPrecursor> for DenunciationIndex {
...
+            (
+                DenunciationIndex::Endorsement { slot, index },
+                DenunciationIndex::Endorsement {
+                    slot: other_slot,
+                    index: other_index,
+                },
+            ) => slot.cmp(other_slot).then(index.cmp(other_index)),
+            (self_de, other_de) => {
+                // Note: does not specify an order for Block header de idx versus Endorsement de idx
+                //       but this should not be an issue
```
[the remark this answers is not in the document]
Comment (sydhds) [reply massa-models/src/denunciation.rs]: Fixed

File: massa-execution-worker/src/speculative_executed_denunciations.rs [hunk 6]
```diff
@@ -65,7 +65,11 @@ impl SpeculativeExecutedDenunciations {
         }
```
Comment (damip) [inline massa-execution-worker/src/speculative_executed_denunciations.rs:65]: is_de_processed should be renamed is_denunciation_executed

**Comment by damip** [comment]
The method "process_denunciation" should be renamed "execute_denunciation"

File: massa-execution-worker/src/speculative_executed_denunciations.rs (hunk 6 above)
Comment (sydhds) [reply massa-execution-worker/src/speculative_executed_denunciations.rs:65]: Done

File: massa-pool-worker/src/denunciation_pool.rs [hunk 7]
```diff
@@ -177,15 +186,32 @@ impl DenunciationPool {
...
+            if let DenunciationStatus::DenunciationEmitted(de) = de_status {
+                // Checks
+                // 1. the denunciation has not been executed already
+                // 2. Denounced item slot is equal or before target slot of block header
+                // 3. Denounced item slot is not too old
+                let de_slot = de.get_slot();
+                if self.execution_controller.is_denunciation_unexecuted(de_idx)
+                    && de_slot <= target_slot
+                    && de_slot.period
+                        >= target_slot
```
[the remark this answers is not in the document]
Comment (sydhds) [reply massa-pool-worker/src/denunciation_pool.rs]: Done

**Comment by sydhds** [comment]
> The method "process_denunciation" should be renamed "execute_denunciation"
> 
> "is_denunciation_processed" should also be renamed

done

File: massa-models/src/denunciation.rs [hunk 8]
```diff
@@ -797,6 +820,28 @@ impl From<&DenunciationPrecursor> for DenunciationIndex {
     }
 }
 
+impl Ord for DenunciationIndex {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let self_key = (
+            self.get_slot(),
+            self.get_index().unwrap_or(&0),
+            u32::from(DenunciationIndexTypeId::from(self)),
```
Comment (damip) [inline massa-models/src/denunciation.rs:828]: As mentioned in the previous review, this should be commented to explain why the type ID should be kept as the last element of the tuple

File: massa-models/src/denunciation.rs [hunk 9]
```diff
@@ -257,6 +258,18 @@ impl Denunciation {
...
+    ) -> bool {
+        // If the Slot is final, a Denunciation can still be made for 1 cycle
+        final_slot_period.checked_sub(*denunciation_slot_period)
+            > Some(*denunciation_expire_periods)
+    }
+
+    /*
     /// For a given slot (and given the slot at now()), check if it can be denounced
     /// Can be used to check if block header | endorsement is not too old (at reception or too cleanup cache)
     pub fn is_expired(
```
Comment (damip) [inline massa-models/src/denunciation.rs:275]: todo remove

File: massa-models/src/denunciation.rs [hunk 10]
```diff
@@ -257,6 +258,18 @@ impl Denunciation {
         }
     }
 
+    /// Check if denunciation has expired given a slot period (assuming the slot is final)
+    pub fn is_expired(
```
Comment (damip) [inline massa-models/src/denunciation.rs:262]: this can be turned into a method of Denunciation

File: massa-models/src/denunciation.rs [hunk 11]
```diff
@@ -257,6 +258,18 @@ impl Denunciation {
         }
     }
 
+    /// Check if denunciation has expired given a slot period (assuming the slot is final)
```
Comment (damip) [inline massa-models/src/denunciation.rs:261]: no need to assume the slot is final, this is an unrelated concept. Is_expired is also used when checking if the denunciation can still be executed when checking a non-final block

File: massa-models/src/denunciation.rs [hunk 12]
```diff
@@ -257,6 +258,18 @@ impl Denunciation {
...
+    pub fn is_expired(
+        denunciation_slot_period: &u64,
+        final_slot_period: &u64,
+        denunciation_expire_periods: &u64,
+    ) -> bool {
+        // If the Slot is final, a Denunciation can still be made for 1 cycle
```
Comment (damip) [inline massa-models/src/denunciation.rs:267]: remove "1 cycle" from the comment: it is a config parameter

File: massa-models/src/denunciation.rs (hunk 8 above)
Comment (sydhds) [reply massa-models/src/denunciation.rs:817]: Added

File: massa-models/src/denunciation.rs (hunk 9 above)
Comment (sydhds) [reply massa-models/src/denunciation.rs]: removed

File: massa-models/src/denunciation.rs (hunk 11 above)
Comment (sydhds) [reply massa-models/src/denunciation.rs]: I've clarified the documentation and mentions the usage so the reader knows when it is used.

File: massa-models/src/denunciation.rs (hunk 12 above)
Comment (sydhds) [reply massa-models/src/denunciation.rs]: done

File: massa-models/src/denunciation.rs (hunk 10 above)
Comment (damip) [reply massa-models/src/denunciation.rs:265]: okay

File: massa-pool-worker/src/denunciation_pool.rs [hunk 13]
```diff
@@ -81,10 +81,11 @@ impl DenunciationPool {
             now,
         );
 
-        if slot.period
-            < self.last_cs_final_periods[slot.thread as usize]
-                .saturating_sub(self.config.denunciation_expire_periods)
-        {
+        if Denunciation::is_expired(
+            &slot.period,
+            &self.last_cs_final_periods[slot.thread as usize],
```
Comment (damip) [inline massa-pool-worker/src/denunciation_pool.rs:86]: Warning: denunciations are not threaded !

Let's update both the pruning and `add_denunciation_precursor` to use `self.last_cs_final_periods.iter().min()` for this check

File: massa-models/src/denunciation.rs [hunk 14]
```diff
@@ -753,9 +754,18 @@ impl DenunciationIndex {
...
+    pub fn get_index(&self) -> Option<&u32> {
+        match self {
+            DenunciationIndex::BlockHeader { .. } => None,
+            DenunciationIndex::Endorsement { slot: _, index } => Some(index),
+        }
+    }
+
     /// Compute the hash
     pub fn get_hash(&self) -> Hash {
         let mut buffer = vec![];
```
Comment (bilboquet) [inline massa-models/src/denunciation.rs:767]: maybe
let mut buffer = u32::from(DenunciationIndexTypeId::from(self)).to_le_bytes().to_vec();

**Comment by damip** [comment]
Security issue: 

Denunciation execution does not check for PoS draws (and older PoS draws might not be available shortly after bootstrap ?).
If I create a block with a fake (but valid) denunciation that targets a slot S with a public key different than that of the person drawn to effectively produce a block at that slot, the denunciation will be executed and prevent any further denunciations of the real culprit.

Let's discuss this.

File: massa-models/src/denunciation.rs (hunk 14 above)
Comment (sydhds) [reply massa-models/src/denunciation.rs]: done

File: massa-pool-worker/src/denunciation_pool.rs (hunk 13 above)
Comment (sydhds) [reply massa-pool-worker/src/denunciation_pool.rs]: done

**Comment by sydhds** [comment]
> Security issue:
> 
> Denunciation execution does not check for PoS draws (and older PoS draws might not be available shortly after bootstrap ?). If I create a block with a fake (but valid) denunciation that targets a slot S with a public key different than that of the person drawn to effectively produce a block at that slot, the denunciation will be executed and prevent any further denunciations of the real culprit.
> 
> Let's discuss this in a followup

Added to issue #3860
